nullaway/src/test/java/com/uber/nullaway/jspecify/GenericMethodTests.java

```diff
@@ -947,6 +951,93 @@ public void arrayCovariantSubtyping() {
+            "    return new SingleColumnRowMapper<>(requiredType);",
+            "  }  ",
+            "  ",
+            "  public <V extends @Nullable Object> List<V> query(RowMapper<V> rowMapper) {",
+            "    throw new RuntimeException();",
+            "  }",
+            "  ",
+            "  public <W> List<@Nullable W> queryForList(Class<W> elementType) {",
+            "    return query(getSingleColumnRowMapper(elementType));",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void selfContainedOptional() {
+    makeHelperWithInferenceFailureWarning()
+        .addSourceLines(
+            "Test.java",
+            "import org.jspecify.annotations.NullMarked;",
+            "import org.jspecify.annotations.Nullable;",
+            "@NullMarked",
+            "public class Test {",
+            "    static class Optional<T> {",
+            "        public static <T> Optional<T> ofNullable(@Nullable T value) {",
+            "            return new Optional<>();",
+            "        }",
+            "        public static <T> Optional<T> of(T value) {",
+            "            throw new RuntimeException();",
+            "        }",
+            "    }",
+            "    public static <U extends @Nullable Object> Optional<U> optionalResultNegative1(@Nullable U value) {",
+            "        return Optional.ofNullable(value);",
+            "    }",
+            "    public static <U> Optional<U> optionalResultNegative2(@Nullable U value) {",
+            "        return Optional.ofNullable(value);",
+            "    }",
+            "    public static <U extends @Nullable Object> Optional<U> optionalResultPositive1(@Nullable U value) {",
+            "        // BUG: Diagnostic contains: Failed to infer type argument nullability",
+            "        return Optional.of(value);",
+            "    }",
+            "    // identical to above, testing the other error message",
+            "    public static <U extends @Nullable Object> Optional<U> optionalResultPositive2(@Nullable U value) {",
+            "        // BUG: Diagnostic contains: passing @Nullable parameter 'value'",
+            "        return Optional.of(value);",
+            "    }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void reassignLocal() {
+    makeHelperWithInferenceFailureWarning()
+        .addSourceLines(
+            "Test.java",
+            "import org.jspecify.annotations.NullMarked;",
+            "import org.jspecify.annotations.Nullable;",
+            "@NullMarked",
+            "public class Test {",
+            "    private <R> @Nullable R make() {",
+            "        return null;",
+            "    }",
+            "    void test() {",
+            "        Object result = null;",
+            "        result = make();",
+            "    }",
+            "}")
+        .doTest();
```

Comment:
Not that it needs docs or anything, but I am curious here... did this fail before this PR in some way?

Reply:
It did fail!  Before, we didn't properly mark when `AssignmentTree`s like `result = make()` were assigning to local variables (we only handled initializers in `VariableTree`s).  So, when inference ran, it thought that the `result` variable had to be `@NonNull`, which then caused inference to fail for the `make()` due to the constraint `@Nullable U <: @NonNull Object`.  Now, we properly detect this is an assignment to a local variable, so we do not generate that constraint (since types of locals are inferred).  I'll add a comment on this one too.